juriscraper/opinions/united_states/administrative_agency/bia.py

```diff
@@ -0,0 +1,107 @@
+
+    def _process_elements(self, elements) -> Dict[str, Any]:
+        """Process the element grouping.
+
+        There is no easy way to parse our the content. Unfortunately, the DOJ
+        admins randomly nest elements and not others. The only consistency
+        is that they content is always split between HR tags.  So we iterate
+        over the elements in order until we find an HR tag and then process
+        the content.  Rinse Wash Repeat.
+
+        Additionally, the only date we have is the year of the decision.
+
+        :param elements: The elements between <hr> tags.
+        :return: Case data
+        """
+        case = {}
+        bold_text = elements[0].xpath(".//strong[1]/.. | .//b[1]/..")
+        if not bold_text:
+            return {}
+        intro_text = (
+            elements[0].xpath(".//strong[1]/.. | .//b[1]/..")[0].text_content()
+        )
+        intro_text = intro_text.replace(";", ",")
+        name, cite = intro_text.split(",", 1)
+        # Unfortunately there are no accessible file dates without PDF parsing
+        # So we generate a date and mark it as date_filed_is_approximate = True
+        case["date_filed_is_approximate"] = True
+        years = re.findall(r"\d{4}", cite)
+        if not years:
+            return {}
+        case["date"] = f"{years[-1]}-07-01"
+        case["status"] = "Unpublished"
+        case["neutral_citation"] = cite
+        case["name"] = name
+        case["url"] = elements[0].xpath(".//a")[0].get("href")
+        case["docket"] = elements[0].xpath(".//a")[0].text_content()
+
+        # Iterate over the P tags that hold the summaries, sometimes
+        summary = []
+        for element in elements:
+            if element.tag == "p":
+                summary.append(element.text_content())
+        case["summary"] = "\n".join(summary).strip()
+        return case
+
+    def _process_html(self):
+        if not self.test_mode_enabled():
+            # Sort the URLS by volume to enable the backscraper
+            # We reverse sort the links by volume and choose the first one
+            # unless we are in a backscraper and then we choose what loop
+            # we are in.
+            if not self.urls:
+                urls = self.html.xpath(
+                    ".//table[1]/tbody/tr/td/a[contains(., 'Volume')]"
+                )
+
+                def get_text(elem):
+                    return elem.text_content()
+
+                self.urls = sorted(urls, key=get_text, reverse=True)
+            self.url = self.urls[self.volume].get("href")
+            # Download the new page of decisions
+            self.html = super()._download()
+
+        # Get the article which will contain all of our content.
+        article = self.html.xpath(".//article")[0]
+        # get the last element in the article
+        # this ends the process_elements method on the final call because no
+        # hr tag is present on the last decision
+        *_, last = article.iter()
```

Comment:
I presume you can't do: `article.iter()[-1]` or `list(article)[-1]`? Doing `*_, last` is a pretty fancy way to just get the last item in a list. We also have a function in CL (and maybe also in JS) called something like `prev_and_next` that gives you something like:

```
for prev, current, next in my_iterable:
    do_something()
```

If you use that you can check for the last item by doing:

```
for prev, current, next in my_iterable:
    if prev is None:
        print("First item")
    if next is None:
        print("last item")
```

It can make some of these types of patterns tidier.

Reply:
I edited this pattern.